ocrd_models/ocrd_models/ocrd_mets.py

```diff
@@ -50,20 +51,91 @@ def empty_mets(now=None):
 
     def __init__(self, **kwargs):
         """
 
         """
         super(OcrdMets, self).__init__(**kwargs)
 
+        # If cache is enabled
+        if self._cache_flag:
+            # Cache for the fileGrps (mets:fileGrp) - a dictionary with Key and Value pair:
+            # Key: 'fileGrp.USE'
+            # Value: a 'fileGrp' object at some memory location 
+            self._fileGrp_cache = {}
+
+            # Cache for the files (mets:file) - two nested dictionaries
+            # The outer dictionary's Key: 'fileGrp.USE'
+            # The outer dictionary's Value: Inner dictionary
+            # The inner dictionary's Key: 'file.ID'
+            # The inner dictionary's Value: a 'file' object at some memory location
+            self._file_cache = {}
+
+            # Note, if the empty_mets() function is used to instantiate OcrdMets
+            # Then the cache is empty even after this operation
+            self._fill_caches()
+
+    def __exit__(self):
+        if self._cache_flag:
+            self._clear_caches()
+
     def __str__(self):
         """
         String representation
         """
         return 'OcrdMets[fileGrps=%s,files=%s]' % (self.file_groups, list(self.find_files()))
 
+    def _fill_caches(self):
+        """
+        Fills the caches with fileGrps and FileIDs
+        """
+
+        tree_root = self._tree.getroot()
+        el_fileGrp_list = tree_root.find(".//mets:fileSec", NS)
+        if el_fileGrp_list is None or len(el_fileGrp_list) == 0:
+            return
+
+        for el_fileGrp in el_fileGrp_list:
+            fileGrp_use = el_fileGrp.get('USE')
+
+            # NOTE: For some reason the el_fileGrp_list contains None values
+            # when testing with the SBB0000F29300010000/data/mets.xml
+            if fileGrp_use is None:
+                continue
+
+            self._fileGrp_cache[fileGrp_use] = el_fileGrp
+            print("_fill_caches> file group added to the cache: %s" % fileGrp_use)
+
+            # Assign an empty dictionary that will hold the files of the added fileGrp
+            self._file_cache[fileGrp_use] = {}
+
+            for el_file in el_fileGrp:
+                file_id = el_file.get('ID')
+                self._file_cache[fileGrp_use].update({file_id : el_file})
+                print("_fill_caches> file added to the cache: %s" % file_id)
+
+        print("_fill_caches> total fileGrp cache elements: %s" % len(self._fileGrp_cache))
+
+    def _clear_caches(self):
+        """
+        Deallocates the caches
+        """
+
+        fileGrp_counter = 0
+
+        for key in list(self._fileGrp_cache):
+            del self._fileGrp_cache[key]
+            fileGrp_counter += 1
+
+        # print("_clear_caches> total cleared fileGrp cache elements: %d" % fileGrp_counter)
+
+        for key in list(self._file_cache):
+            for inner_key in list(self._file_cache[key]):
+                del self._file_cache[key][inner_key]
+            del self._file_cache[key]
```

Comment:
Why not simply `self._fileGrp_cache = None` or `del self._file_cache`? Python will garbage-collect dangling references. But the references will remain anyway, because they are tied to `self._tree`.

Reply:
They should be, GC in Python is fairly robust.

Reply:
Okay.